compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1580,7 +1580,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let coerce = ctxt.coerce.as_mut().unwrap();
             if let Some((tail_expr, tail_expr_ty)) = tail_expr_ty {
                 let span = self.get_expr_coercion_span(tail_expr);
-                let cause = self.cause(span, ObligationCauseCode::BlockTailExpression(blk.hir_id));
+                let cause = self.cause(
+                    span,
+                    ObligationCauseCode::BlockTailExpression(
+                        blk.hir_id,
+                        blk.hir_id,
```

Comment:
I am assuming from context that this second hir id is the scrutinee. It seems like it needs to be an option, otherwise it's kinda just confusing to be passing this hir id twice for this case :sweat_smile:

Reply:
Yeah, I know. It's funny because the scrutinee should really be included in the MatchSource to begin with.

Reply:
Moved it into MatchSource